File: app/src/main/java/com/example/covidnow/fragment/MapsFragment.java
package com.example.covidnow.fragment;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;
import androidx.core.util.Pair;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.Manifest;
import android.content.pm.PackageManager;
import android.location.Location;
import android.os.Bundle;
import android.os.Looper;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Toast;

import com.codepath.asynchttpclient.AsyncHttpClient;
import com.codepath.asynchttpclient.RequestParams;
import com.codepath.asynchttpclient.callback.JsonHttpResponseHandler;
import com.example.covidnow.adapter.PlacesAdapter;
import com.example.covidnow.R;
import com.example.covidnow.repository.PlacesRepository;
import com.example.covidnow.viewmodels.HomeViewModel;
import com.example.covidnow.viewmodels.MapsViewModel;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.LocationSettingsRequest;
import com.google.android.gms.location.SettingsClient;
import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.parse.GetCallback;
import com.parse.ParseException;
import com.parse.ParseQuery;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import okhttp3.Headers;
import permissions.dispatcher.NeedsPermission;
import permissions.dispatcher.RuntimePermissions;

import static com.google.android.gms.location.LocationServices.getFusedLocationProviderClient;

@RuntimePermissions
public class MapsFragment extends Fragment {

    private static final String TAG = "MapsFragment";
    private MapsFragment fragment = this;
    private EditText etSearch;
    private ImageButton btnSearch;
    private RecyclerView rvPlaces;
    private final static String KEY_LOCATION = "location";
    private LocationRequest mLocationRequest;
    private long UPDATE_INTERVAL = 60000;  /* 60 secs */
    private long FASTEST_INTERVAL = 5000; /* 5 secs */
    private static List<com.example.covidnow.models.Location> adapterPlaces = new ArrayList<>();
    private static PlacesAdapter adapter;
    private Pair<Double, Double> coordinates;
    private Location mCurrentLocation;
    private GoogleMap map;
    private MapsViewModel mViewModel;

    private OnMapReadyCallback callback = new OnMapReadyCallback() {

        /**
         * Manipulates the map once available.
         * This callback is triggered when the map is ready to be used.
         * This is where we can add markers or lines, add listeners or move the camera.
         * If Google Play services is not installed on the device, the user will be prompted to
         * install it inside the SupportMapFragment. This method will only be triggered once the
         * user has installed Google Play services and returned to the app.
         */
        @Override
        public void onMapReady(GoogleMap googleMap) {
            map = googleMap;

            if (map != null) {
                // Map is ready
                // Get users current location with permission
                MapsFragmentPermissionsDispatcher.getMyLocationWithPermissionCheck(fragment);
                MapsFragmentPermissionsDispatcher.startLocationUpdatesWithPermissionCheck(fragment);
            } else {
                Toast.makeText(fragment.getContext(), "Error - Map was null!!", Toast.LENGTH_SHORT).show();
            }
        }
    };

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_maps, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (savedInstanceState != null && savedInstanceState.keySet().contains(KEY_LOCATION)) {
            // Since KEY_LOCATION was found in the Bundle, we can be sure that mCurrentLocation
            // is not null.
            Log.i(TAG, "mCurrentLocation not null");
            mCurrentLocation = savedInstanceState.getParcelable(KEY_LOCATION);
        }
        // Set view model
        mViewModel = ViewModelProviders.of(this).get(MapsViewModel.class);

        etSearch = view.findViewById(R.id.etSearch);
        btnSearch = view.findViewById(R.id.btnSearch);
        rvPlaces = view.findViewById(R.id.rvPlaces);

        // Initialize adapter
        adapterPlaces = new ArrayList<>();
        adapter = new PlacesAdapter(fragment, adapterPlaces);


        SupportMapFragment mapFragment =
                (SupportMapFragment) getChildFragmentManager().findFragmentById(R.id.map);

        if (mapFragment != null) {
            mapFragment.getMapAsync(callback);
        }

        // Listen for searches
        btnSearch.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String search = etSearch.getText().toString();
                if (search.isEmpty()) {
                    Toast.makeText(getContext(), "Must provide search query", Toast.LENGTH_SHORT).show();
                } else {
                    rvPlaces.setAdapter(adapter);
                    // Locate nearby places
                    if (coordinates == null) {
                        Toast.makeText(getContext(), "Error, current location not found yet", Toast.LENGTH_SHORT).show();
                        return;
                    }
                    mViewModel.getPlaces(coordinates, search, getString(R.string.google_maps_key), getViewLifecycleOwner());

                    // Listen for nearby places JSON from PlacesRepository
                    final Observer<JSONArray> placesJSONObserver = new Observer<JSONArray>() {
                        @Override
                        public void onChanged(@Nullable final JSONArray jArray) {
                            // Location is ready to be passed to Places API
                            Log.i(TAG, "Places JSON received from PlacesRepo");
                            mViewModel.getSavedPlaces(jArray);
                        }
                    };
                    mViewModel.getNearbyPlacesJson().observe(getViewLifecycleOwner(), placesJSONObserver);

                    // Make recyclerview visible at the bottom
                    rvPlaces.setVisibility(View.VISIBLE);
                    // Initialize recyclerview

                    LinearLayoutManager layoutManager = new LinearLayoutManager(getContext());
                    rvPlaces.setLayoutManager(layoutManager);

                    // Add lines between recycler view
                    RecyclerView.ItemDecoration itemDecoration = new DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL);
                    rvPlaces.addItemDecoration(itemDecoration);

                    // Listen for List<Location> received from ParseRepo
                    final Observer<List<com.example.covidnow.models.Location>> placesListObserver = new Observer<List<com.example.covidnow.models.Location>>() {
                        @Override
                        public void onChanged(@Nullable final List<com.example.covidnow.models.Location> newPlaces) {
                            // List of places ready to be given to recyclerview
                            Log.i(TAG, "Places list received from ParseRepo");
                            adapter.addAll(newPlaces);
                        }
                    };
                    mViewModel.getNearbyPlacesList().observe(getViewLifecycleOwner(), placesListObserver);
                }
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        MapsFragmentPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
    }

    @SuppressWarnings({"MissingPermission"})
    @NeedsPermission({Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION})
    public void getMyLocation() {
        map.setMyLocationEnabled(true);
        map.getUiSettings().setMyLocationButtonEnabled(true);
        // Access users current location
        Log.i(TAG, "Context: " + fragment.getContext());
        FusedLocationProviderClient locationClient = getFusedLocationProviderClient(fragment.getContext());
        locationClient.getLastLocation()
                .addOnSuccessListener(new OnSuccessListener<Location>() {
                    @Override
                    public void onSuccess(Location location2) {
                        if (location2 != null) {
                            Log.i(TAG, "Location: " + location2.toString());
                            // Give coordinates to view model
                            coordinates = Pair.create(location2.getLatitude(), location2.getLongitude());
                            mCurrentLocation = location2;
                            moveCameraToLocation(location2);
                            // Listen for location changes
                            onLocationChanged(location2);
                        }
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Log.d(TAG, "Error trying to get last GPS location");
                        e.printStackTrace();
                    }
                });
    }

    public void onSaveInstanceState(Bundle savedInstanceState) {
        savedInstanceState.putParcelable(KEY_LOCATION, mCurrentLocation);
        super.onSaveInstanceState(savedInstanceState);
    }

    private void moveCameraToLocation(Location loc) {
        double x = loc.getLatitude();
        double y = loc.getLongitude();
        // Set camera to user's current location
        LatLng currLocationLatLng = new LatLng(x, y);
        CameraUpdate cameraUpdate = CameraUpdateFactory.newLatLngZoom(currLocationLatLng, 17);
        map.animateCamera(cameraUpdate);

    }

    /*
     * Called when the Activity becomes visible.
     */
    @Override
    public void onStart() {
        super.onStart();
    }

    /*
     * Called when the Activity is no longer visible.
     */
    @Override
    public void onStop() {
        super.onStop();
    }

    @NeedsPermission({Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION})
    protected void startLocationUpdates() {
        mLocationRequest = new LocationRequest();
        mLocationRequest.setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY);
        mLocationRequest.setInterval(UPDATE_INTERVAL);
        mLocationRequest.setFastestInterval(FASTEST_INTERVAL);

        LocationSettingsRequest.Builder builder = new LocationSettingsRequest.Builder();
        builder.addLocationRequest(mLocationRequest);
        LocationSettingsRequest locationSettingsRequest = builder.build();

        SettingsClient settingsClient = LocationServices.getSettingsClient(Objects.requireNonNull(getContext()));
        settingsClient.checkLocationSettings(locationSettingsRequest);
        //noinspection MissingPermission
        if (ActivityCompat.checkSelfPermission(Objects.requireNonNull(getContext()), Manifest.permission.ACCESS_FINE_LOCATION)
                != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(getContext(),
                Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            // TODO: Consider calling
            //    ActivityCompat#requestPermissions
            // here to request the missing permissions, and then overriding
            //   public void onRequestPermissionsResult(int requestCode, String[] permissions,
            //                                          int[] grantResults)
            // to handle the case where the user grants the permission. See the documentation
            // for ActivityCompat#requestPermissions for more details.
            return;
        }
        getFusedLocationProviderClient(getContext()).requestLocationUpdates(mLocationRequest, new LocationCallback() {
                    @Override
                    public void onLocationResult(LocationResult locationResult) {
                        onLocationChanged(locationResult.getLastLocation());
                    }
                },
                Looper.myLooper());
    }

    public void onLocationChanged(Location location2) {
        // GPS may be turned off
        if (location2 == null) {
            return;
        }
        // Report to the UI that the location was updated
        mCurrentLocation = location2;
    }
}

File: app/src/main/java/com/example/covidnow/fragment/MapsFragment.kt
package com.example.covidnow.fragment

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Canvas
import android.graphics.drawable.ColorDrawable
import android.location.Location
import android.os.Build
import android.os.Bundle
import android.os.Looper
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.view.inputmethod.InputMethodManager
import android.widget.*
import androidx.annotation.RequiresApi
import androidx.cardview.widget.CardView
import androidx.core.app.ActivityCompat
import androidx.core.util.Pair
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import com.example.covidnow.R
import com.example.covidnow.adapter.PlacesAdapter
import com.example.covidnow.helpers.RecyclerViewSwipeListener
import com.example.covidnow.viewmodels.MapsViewModel
import com.google.android.gms.location.*
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.material.floatingactionbutton.FloatingActionButton
import org.json.JSONArray
import org.parceler.Parcels
import permissions.dispatcher.NeedsPermission
import permissions.dispatcher.RuntimePermissions
import java.util.*

@RuntimePermissions
class MapsFragment : Fragment() {
    private val fragment = this
    private var etSearch: EditText? = null
    private var btnSearch: ImageButton? = null
    private var btnQuickReview: FloatingActionButton? = null
    private var card: CardView? = null
    private var rvPlaces: RecyclerView? = null
    private var ivArrow: ImageView? = null
    private var coordinates: Pair<Double, Double>? = null
    private var pbLoading: ProgressBar? = null
    private var mCurrentLocation: Location? = null
    private var map: GoogleMap? = null
    private var mViewModel: MapsViewModel? = null
    private val permissionFineLocation = Manifest.permission.ACCESS_FINE_LOCATION
    private val permissionCoarseLocation = Manifest.permission.ACCESS_COARSE_LOCATION
    private val REQUEST_CODE_LOCATION = 100
    private val callback = OnMapReadyCallback { googleMap ->
        /**
         * Manipulates the map once available.
         * This callback is triggered when the map is ready to be used.
         */
        map = googleMap
        if (map != null) {
            // Map is ready
            // Get users current location with permission
            if (validatePermissionsLocation()){
                Log.i(TAG, "permission")
                getMyLocation()
            }
            else{
                requestPermissions()
            }
        } else {
            Toast.makeText(fragment.context, "Error - Map was null!", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_maps, container, false)
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        pbLoading = view.findViewById(R.id.pbLoading)
        etSearch = view.findViewById(R.id.etSearch)
        btnSearch = view.findViewById(R.id.btnSearch)
        rvPlaces = view.findViewById(R.id.rvPlaces)
        card = view.findViewById(R.id.card)
        btnQuickReview = view.findViewById(R.id.btnQuickReview)
        ivArrow = view.findViewById(R.id.ivArrow)

        var flag = true

        if (validatePermissionsLocation()){
            Log.i(TAG, "Permission granted")
            getMyLocation()
        }
        else{
            Log.i(TAG, "Permission not granted")
            //requestPermissions()
        }

        if (savedInstanceState != null && savedInstanceState.keySet().contains(KEY_LOCATION)) {
            // Since KEY_LOCATION was found in the Bundle, we can be sure that mCurrentLocation
            // is not null.
            Log.i(TAG, "mCurrentLocation not null")
            mCurrentLocation = savedInstanceState.getParcelable(KEY_LOCATION)
        }
        // Set view model
        mViewModel = ViewModelProviders.of(this).get(MapsViewModel::class.java)

        // Setup recyclerview of places
        initializeRvPlaces(rvPlaces)

        // Setup map view
        val mapFragment = childFragmentManager.findFragmentById(R.id.map) as SupportMapFragment?
        mapFragment?.getMapAsync(callback)

        // Listen for location to be retrieved for quick review
        val coordinatesObserver: Observer<Pair<Double, Double>> = Observer { it ->
            Log.i(TAG, "Coordinates received from view model")
            // Move camera to current location
            coordinates = it
            it.first?.let { it1 -> it.second?.let { it2 -> moveCameraToCoordinates(it1, it2) } }
        }
        mViewModel?.getCoordinates()?.observe(viewLifecycleOwner, coordinatesObserver)

        // Listen for searches
        btnSearch?.setOnClickListener(View.OnClickListener {
            // Show progress bar
            if (coordinates == null) {
                Toast.makeText(context, "Error, current location not found yet", Toast.LENGTH_SHORT).show()
                return@OnClickListener
            }
            // Get user's query
            val search = etSearch?.text.toString()
            // Clear search box
            etSearch?.setText("")
            if (search.isEmpty()) {
                Toast.makeText(context, "Must provide search query", Toast.LENGTH_SHORT).show()
            } else {
                // Automatically put keyboard away
                val mgr = activity?.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                mgr.hideSoftInputFromWindow(etSearch?.windowToken, 0)

                // Show progress bar
                pbLoading?.visibility = View.VISIBLE

                // Retrieve nearby places
                if (coordinates != null) {
                    mViewModel?.getPlaces(coordinates as Pair<Double, Double>, search, getString(R.string.google_maps_key))
                }

                // Listen for nearby places JSON from PlacesRepository
                val placesJSONObserver: Observer<JSONArray> = Observer<JSONArray> { jArray -> // Places JSON received from view model
                    Log.i(TAG, "Places JSON received from PlacesRepo")
                    mViewModel?.getSavedPlaces(jArray as JSONArray)
                }
                // Listen for Places API call to finish
                mViewModel?.getNearbyPlacesJson()?.observe(viewLifecycleOwner, placesJSONObserver)
                val placesListObserver: Observer<List<com.example.covidnow.models.Location>> = Observer<List<com.example.covidnow.models.Location>> { newPlaces -> // List of places ready to be given to recyclerview
                    Log.i(TAG, "Places list received from ParseRepo")
                    adapter?.addAll(newPlaces as MutableList<com.example.covidnow.models.Location>)
                    // Hide progress bar
                    pbLoading?.visibility = View.GONE
                    // Make places list slide up
                    if (flag) {
                        showPlaces()
                    }
                    flag = false

                }
                // Listen for List<covidnow.location> of saved places received from ParseRepo
                mViewModel?.getNearbyPlacesList()?.observe(viewLifecycleOwner, placesListObserver)
            }
        })

        // Listen for arrow button to show rvPlaces
        ivArrow?.setOnClickListener(View.OnClickListener { showPlaces() })

        // Listen for location to be retrieved for quick review
        val finalLocationObserver: Observer<com.example.covidnow.models.Location> = Observer<com.example.covidnow.models.Location> {
            Log.i(TAG, "Location received from view model as Location")
            // Listen for user to press quick review button
            btnQuickReview?.setOnClickListener(View.OnClickListener {
                Log.i(TAG, "Quick Review button clicked!")
                // Location ready to be used for quick review
                goQuickReview(mViewModel?.getFinalLocation()?.value)
            })
        }
        // Listen for location to be ready to be reviewed
        mViewModel?.getFinalLocation()?.observe(viewLifecycleOwner, finalLocationObserver)
    }

    private fun initializeRvPlaces(rvPlaces: RecyclerView?) {
        // Initialize adapter
        adapterPlaces = ArrayList()
        adapter = PlacesAdapter(fragment, adapterPlaces as ArrayList<com.example.covidnow.models.Location>)

        // Initialize recyclerview
        rvPlaces?.adapter = adapter
        val layoutManager = LinearLayoutManager(context)
        rvPlaces?.layoutManager = layoutManager

        // Add lines between recycler view
        val itemDecoration: ItemDecoration = DividerItemDecoration(context, DividerItemDecoration.VERTICAL)
        rvPlaces?.addItemDecoration(itemDecoration)

        // Initialize swipe for details on rvPlaces
        val touchHelperCallback: ItemTouchHelper.SimpleCallback = object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {
            private val background = ColorDrawable(resources.getColor(R.color.light_blue))
            override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
                return false
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                adapter?.goToDetails(viewHolder.adapterPosition)
            }

            // Show blue background on swipe
            override fun onChildDraw(c: Canvas, recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, dX: Float, dY: Float, actionState: Int, isCurrentlyActive: Boolean) {
                super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
                val itemView = viewHolder.itemView
                if (dX > 0) {
                    background.setBounds(itemView.left, itemView.top, itemView.left + dX.toInt(), itemView.bottom)
                } else if (dX < 0) {
                    background.setBounds(itemView.right + dX.toInt(), itemView.top, itemView.right, itemView.bottom)
                } else {
                    background.setBounds(0, 0, 0, 0)
                }
                background.draw(c)
            }
        }
        val itemTouchHelper = ItemTouchHelper(touchHelperCallback)
        itemTouchHelper.attachToRecyclerView(rvPlaces)

        // Swipe up to show places, swipe down to hide
        rvPlaces?.onFlingListener = object : RecyclerViewSwipeListener(true) {
            override fun onSwipeDown() {
                hidePlaces()
            }

            override fun onSwipeUp() {
                //showPlaces()
            }
        }
    }

    private fun goQuickReview(location: com.example.covidnow.models.Location?) {
        val newFrag: Fragment = ComposeReviewFragment()
        val result = Bundle()
        // Send this location to the compose fragment
        result.putParcelable("location", Parcels.wrap(location))
        newFrag.arguments = result
        // Start compose review fragment
        fragmentManager?.beginTransaction()?.replace(R.id.flContainer,
                newFrag)?.addToBackStack("MapsFragment")?.commit()
    }

    private fun showPlaces() {
        // Show rvPlaces with animation
        Log.i(TAG, "Showing places")
        card?.visibility = View.VISIBLE
        card?.startAnimation(AnimationUtils.loadAnimation(context,
                R.anim.slide_up))
        ivArrow?.visibility = View.GONE
    }

    fun addMarker(location: com.example.covidnow.models.Location) {
        // Add marker at current location
        val point = LatLng(location.latitude, location.longitude)
        val marker = map?.addMarker(mViewModel?.createMarker(point, location))
        // Move camera to new marker
        moveCameraToLatLng(point)
        marker?.let { mViewModel?.dropPinEffect(it) }
    }

    private fun hidePlaces() {
        // Hide rvPlaces with animation
        Log.i(TAG, "Hiding places")
        card?.visibility = View.GONE
        card?.startAnimation(AnimationUtils.loadAnimation(context,
                R.anim.slide_down))
        ivArrow?.visibility = View.VISIBLE
    }

    private fun requestPermissions(){
        val contextProvider= activity?.let { ActivityCompat.shouldShowRequestPermissionRationale(it, permissionFineLocation) }

        if(contextProvider == true){
            Toast.makeText(activity?.applicationContext, "Permission is required to obtain location", Toast.LENGTH_SHORT).show()
        }
        permissionRequest()
    }
    private fun permissionRequest(){
        activity?.let { ActivityCompat.requestPermissions(it, arrayOf(permissionFineLocation, permissionCoarseLocation), REQUEST_CODE_LOCATION) }
    }
    private fun validatePermissionsLocation():Boolean{
        val fineLocationAvailable= activity?.applicationContext?.let { ActivityCompat.checkSelfPermission(it, permissionFineLocation) } == PackageManager.PERMISSION_GRANTED
        val coarseLocationAvailable= activity?.applicationContext?.let { ActivityCompat.checkSelfPermission(it, permissionCoarseLocation) } ==PackageManager.PERMISSION_GRANTED

        return fineLocationAvailable && coarseLocationAvailable
    }

    @NeedsPermission(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION)
    fun getMyLocation() {
        Log.i(TAG, "GETTING  LIOCATION ")
        if (context?.let { ActivityCompat.checkSelfPermission(it, Manifest.permission.ACCESS_FINE_LOCATION) } != PackageManager.PERMISSION_GRANTED && context?.let { ActivityCompat.checkSelfPermission(it, Manifest.permission.ACCESS_COARSE_LOCATION) } != PackageManager.PERMISSION_GRANTED) {
            permissionRequest()
            return
        }
        map?.isMyLocationEnabled = true
        map?.uiSettings?.isMyLocationButtonEnabled = true
        val locationClient = activity?.applicationContext?.let { LocationServices.getFusedLocationProviderClient(it) }
        Log.i(TAG, "LOCATION CLIENT " + locationClient)
        mViewModel?.getMyLocation(locationClient, getString(R.string.google_maps_key))
    }

    private fun moveCameraToCoordinates(x: Double, y: Double) {
        // Set camera to user's current location
        val currLocationLatLng = LatLng(x, y)
        val cameraUpdate = CameraUpdateFactory.newLatLngZoom(currLocationLatLng, 17f)
        map?.animateCamera(cameraUpdate)
    }

    private fun moveCameraToLatLng(point: LatLng) {
        // Set camera to this point
        val cameraUpdate = CameraUpdateFactory.newLatLngZoom(point, 17f)
        map?.animateCamera(cameraUpdate)
    }

    companion object {
        private const val TAG = "MapsFragment"
        private const val KEY_LOCATION = "location"
        private var adapterPlaces: List<com.example.covidnow.models.Location> = ArrayList()
        private var adapter: PlacesAdapter? = null
    }
}